cli/fs.rs

```diff
@@ -49,8 +49,12 @@ pub fn write_file_2<T: AsRef<[u8]>>(
 
 #[cfg(unix)]
 fn set_permissions(file: &mut File, mode: u32) -> std::io::Result<()> {
-  debug!("set file mode to {}", mode);
-  file.set_permissions(PermissionsExt::from_mode(mode & 0o777))
+  let mode = mode & 0o777;
+  debug!("set file mode to {:o}", mode);
+  let metadata = file.metadata()?;
+  let mut permissions = metadata.permissions();
+  permissions.set_mode(mode);
+  file.set_permissions(permissions)
```

Comment:
Why is this better? It looks like might make more syscalls.

Reply:
As I understand the Rust docs, the previous behavior was buggy. If a file was marked as "read-only" (this is exposed in the platform-neutral std::fs API, but I'm not sure what it amounts to on Unix, and I have minimal Windows experience,...) and it's run through the previous implementation of this function, then those read-only bits get erased and instead the `mode` bits are written to the file with no additional metadata. Same if the file has any metadata that the Rust API doesn't now expose. However, we can't make a test for this in TypeScript because we don't now have an API for creating or querying such additional metadata.

Reply:
Check out [std::fs::Permissions](https://doc.rust-lang.org/std/fs/struct.Permissions.html) and [std::os::unix::fs::PermissionsExt](https://doc.rust-lang.org/std/os/unix/fs/trait.PermissionsExt.html). The old method creates a new permissions structure containing only the desired `mode` bits. The new method proposed here (and at the second link) reads the existing permissions structure, overwrites just its `mode` bits with the desired new ones, and then writes that permissions structure back to the file.

Reply:
I don't think there's a bug here. This isn't an op, just an internal function so it can be tested with a Rust unit test. If you can demonstrate a bug I would happily accept it. As this is now, as far as I can see, it's just moving code around. Thank you for the effort but I'm closing this PR without merge.

Reply:
Right, forgot we can make a Rust unit test. I'll see if I can demonstrate a bug doing that.

Reply:
@ry, it looks like you were correct. On Unix, the `permissions().readonly`/`set_readonly` only affect the chmod bits, and on Windows, the `set_permissions` function is a noop anyway, so it doesn't interfere with any readonly bits that may have been externally applied to a file. So the "fix" I proposed was unnecessary. Sorry for the noise.